accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,253 @@
+
+This is the technical description of the experience described in [Trusted Publishers, using OpenID Connect for NuGet
+push](trusted-publishers-oidc-for-nuget-push.md). The content is split to clearly separate the functional/UX description
+(the other doc) from the technical description (this doc). This document acts as a supporting document to the other so
+that other one should be read first.
+
+Much of the technical explanation is described in [Trusted Publishers for All Package
+Repositories](https://repos.openssf.org/trusted-publishers-for-all-package-repositories). I will expand on certain
+technical details which are particularly interesting or specific to NuGet.
+
+## Validation of the GitHub OIDC token
+
+The most critical step in this design is understanding how NuGet.org will validate that an incoming OIDC token is
+acceptable to trade for a short-lived NuGet API key.
+
+The following checks are made:
+
+- The token is issued by a known Trusted Publisher (only GitHub Actions at this time)
+- The token is validated per all JWT rules, such as but not limited to:
+  - valid signature via JWKS
+  - validate duration (`nbf` and `exp` claims)
+  - only used once, via `jti`
+  - valid `aud` claim, being `nuget`
+
+The
+[Microsoft.IdentityModel.Protocols.OpenIdConnect](https://www.nuget.org/packages/Microsoft.IdentityModel.Protocols.OpenIdConnect)
+package can help us properly validate JWTs in our NuGet/NuGetGallery ASP.NET application.
+
+In addition to the general JWT checks, specific checks are made for each Trusted Publisher. For GitHub Actions, the
+following checks will be made:
+
+- `sub` claim has a `repo:{repo owner}/{repo name}:` prefix (case insensitive)
+  - The suffix of the `sub` is implied by the other checks
+  - Example, when an environment is set: `repo:octo-org/octo-repo:environment:Production`
+  - Example, for a specific branch and no environment is set: `repo:octo-org/octo-repo:ref:refs/heads/demo-branch`
+- `repository_owner` claim matches the `{repo owner}` name (case insensitive)
+- `repository_owner_id` claim matches the numeric owner ID recorded at the time of the trust policy creation
+  - This is to avoid resurrection attacks.
+- `repository` claim matches the `{repo name}` name (case insensitive)
+- `repository_id` claim matches the numeric repository ID recorded at the time of the trust policy creation
+  - This is to avoid resurrection attacks.
+- If a branch filter is provided in the trust policy:
+  - `ref_type` claim must be `branch`
+  - `ref` claim must be `refs/head/{branch}` (case sensitive, branch names of differing case can coexist, branch match supports wildcards)
+- If an environment filter is provided in the trust policy:
+  - `environment` claim must be `{environment}` (case insensitive)
+- If a workflow path filter is provided in the trust policy:
+   - `job_workflow_ref` claim must be `{repo owner}/{repo name}/{workflow path}@.*` (case insensitive)
```

Comment:
I _think_ this should be case sensitive? The workflow path at least is definitely case sensitive.

Reply:
I chose to relax case sensitivity here due to Windows being a case insensitive file system. I don't feel too strongly, but it seems like a silly reason to reject a token JUST because the case was wrong, assuming everything else matches.

Reply:
Since GitHub is case sensitive, it does mean that more than one workflow could gain the ability to publish. In practice I think that would be fairly uncommon (probably mostly indicative of a typo) but might be a surprising side effect.

Reply:
Yes, that could be weird. I am not sure if GH has mitigations in place for multiple files of the same case. I'll listen for other input on this, and we can make a call during implementation.